js/src/Ice/OutgoingAsync.js

```diff
@@ -151,16 +145,148 @@ export class ProxyOutgoingAsyncBase extends OutgoingAsyncBase {
+                // router). In this case, we must *always* retry, so that the missing proxy is added to the router.
+                ref.getRouterInfo().clearCache(ref);
+
+                if (traceLevels.retry >= 1) {
+                    logger.trace(traceLevels.retryCat, "retrying operation call to add proxy to router\n" + ex);
+                }
+                // We must always retry, so we don't look at the retry count.
+                return 0;
+            } else if (ref.isIndirect()) {
+                // We retry ObjectNotExistException if the reference is indirect.
+                if (ref.isWellKnown()) {
+                    const li = ref.getLocatorInfo();
+                    if (li !== null) {
+                        li.clearCache(ref);
+                    }
+                }
+            } else {
+                // For all other cases, we don't retry ObjectNotExistException.
+                throw ex;
+            }
+        } else if (ex instanceof RequestFailedException) {
+            throw ex;
+        }
+
+        // There is no point in retrying an operation that resulted in a MarshalException. This must have been raised
+        // locally (because if it happened in a server it would result in an UnknownLocalException  instead), which
+        // means there was a problem in this process that will not change if we try again.
+        //
+        // The most likely cause for a MarshalException is exceeding the maximum message size, which is represented by
+        // the subclass MemoryLimitException. For example, a client can attempt to send a message that exceeds the
+        // maximum memory size, or accumulate enough batch requests without flushing before the maximum size is reached.
+        //
+        // This latter case is especially problematic, because if we were to retry a batch request after a
+        // MarshalException, we would in fact silently discard the accumulated requests and allow new batch requests to
+        // accumulate. If the subsequent batched requests do not exceed the maximum message size, it appears to the
+        // client that all of the batched requests were accepted, when in reality only the last few are actually sent.
+        if (ex instanceof MarshalException) {
+            throw ex;
+        }
+
+        // Don't retry if the communicator is destroyed, object adapter is deactivated, or connection is manually closed.
+        if (
+            ex instanceof CommunicatorDestroyedException ||
+            ex instanceof ObjectAdapterDeactivatedException ||
+            ex instanceof ConnectionManuallyClosedException
+        ) {
+            throw ex;
+        }
+
+        // Don't retry invocation timeouts.
+        if (ex instanceof InvocationTimeoutException || ex instanceof InvocationCanceledException) {
+            throw ex;
+        }
+
+        ++this._cnt.value;
+        Debug.assert(this._cnt.value > 0);
```

Comment:
you should remove `.value` now that we no longer wrap the _cnt in an object.

Reply:
Another good catch... Fixed.